src/clients/FtpClient.cc

```diff
@@ -763,6 +763,8 @@ Ftp::Client::connectDataChannel()
     conn->remote.port(data.port);
     conn->tos = ctrl.conn->tos;
     conn->nfmark = ctrl.conn->nfmark;
+    // Using non-local addresses in TPROXY mode requires appropriate socket option.
+    conn->flags = ctrl.conn->flags & COMM_TRANSPARENT;
```

Comment:
We should not be overwriting the entire `conn->flags` value like this. If you want to copy the `COMM_TRANSPARENT` bit from the control connection flags, please copy it without touching the other bits of the `conn->flags`.

Please note that the above would remain true even if `conn->flags` were always zero at the time of assignment in the current code (but it is not zero AFAICT).

Reply:
Done